File: CNC modeling 2.0/src/main/java/com/sergey/pisarev/model/ResizableCanvas.java
package com.sergey.pisarev.model;


import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.paint.Color;

public class ResizableCanvas extends Canvas {
    GraphicsContext gc = getGraphicsContext2D();
    int canvasWidth = 0;
    int canvasHeight = 0;
    int halfCanvasHeight = 0;

    /**
     * Constructor
     */
    public ResizableCanvas() {
        // if i didn't add the draw to the @Override resize(double width, double
        // height) then it must be into the below listeners

        // Redraw canvas when size changes.
        widthProperty().addListener((observable, oldValue, newValue) -> {
            System.out.println("Entered WIDTH property");
            canvasWidth = (int) widthProperty().get();
        });
        heightProperty().addListener((observable, oldValue, newValue) -> {
            System.out.println("Entered HEIGHT property");

            canvasHeight = (int) heightProperty().get();
            halfCanvasHeight = canvasHeight >> 1;
        });

    }

    @Override
    public double minHeight(double width) {
        return 1;
    }

    @Override
    public double maxHeight(double width) {
        return Double.MAX_VALUE;
    }

    @Override
    public double prefHeight(double width) {
        return minHeight(width);
    }

    @Override
    public double minWidth(double height) {
        return 1;
    }

    @Override
    public double maxWidth(double height) {
        return Double.MAX_VALUE;
    }

    @Override
    public boolean isResizable() {
        return true;
    }

    @Override
    public void resize(double width, double height) {
        super.setWidth(width);
        super.setHeight(height);
    }
}

File: CNC modeling 2.0/src/main/java/com/sergey/pisarev/model/ResizableCanvas.kt
package com.sergey.pisarev.model

import javafx.beans.value.ObservableValue
import javafx.scene.canvas.Canvas

class ResizableCanvas : Canvas() {
    var gc = graphicsContext2D
    var canvasWidth = 0
    var canvasHeight = 0
    var halfCanvasHeight = 0
    override fun minHeight(width: Double): Double {
        return 1.0
    }

    override fun maxHeight(width: Double): Double {
        return Double.MAX_VALUE
    }

    override fun prefHeight(width: Double): Double {
        return minHeight(width)
    }

    override fun minWidth(height: Double): Double {
        return 1.0
    }

    override fun maxWidth(height: Double): Double {
        return Double.MAX_VALUE
    }

    override fun isResizable(): Boolean {
        return true
    }

    override fun resize(width: Double, height: Double) {
        super.setWidth(width)
        super.setHeight(height)
    }

    /**
     * Constructor
     */
    init {
        // if i didn't add the draw to the @Override resize(double width, double
        // height) then it must be into the below listeners

        // Redraw canvas when size changes.
        widthProperty().addListener { observable: ObservableValue<out Number?>?, oldValue: Number?, newValue: Number? ->
            //System.out.println("Entered WIDTH property");
            canvasWidth = widthProperty().get().toInt()
        }
        heightProperty().addListener { observable: ObservableValue<out Number?>?, oldValue: Number?, newValue: Number? ->
            //System.out.println("Entered HEIGHT property");
            canvasHeight = heightProperty().get().toInt()
            halfCanvasHeight = canvasHeight shr 1
        }
    }
}